### ISSUE TITLE
Discuss Proposed - 0024-crust-refactor

### ISSUE THREAD

**Issue by rossmuir** [maidsafe/rfcs#107]
https://github.com/maidsafe/rfcs/tree/master/proposed/0024-crust-refactor

**Comment by dirvine** [comment]
@canndrew Continued form slack 
######################
Is it that listeners may do a 2 stage connect either plain (if we are seed node) or anonymous encrypted (if we are normal bootstrap ​*or*​ manual port mapped/upnp mapped).

I think encryption at very lowest level is good as we can encrypt any protocol changes over tcp / udp (such as utp) hiding even the fact we use udt / utp etc.

So on listeners where do do not know if it is bootstrap (so one step) or an actual node connect we may end up with several connections per node again. Whereas on rendezvous we would not parse early messages or incorrect peers which is then single connection per peers ?

We may need to check on listener connect if we are expecting this connection and also when we get an expect_connection call from crust then we should also look at existing connections to see if that node is already there.

**Comment by canndrew** [comment]
> Is it that listeners may do a 2 stage connect either plain (if we are seed node) or anonymous encrypted (if we are normal bootstrap ​or​ manual port mapped/upnp mapped).

Why do we need unencrypted connections? We could always distribute the public keys of the seed nodes along with their addresses, and without it someone could MITM and impersonate a seed node. I just think that if we're going to do encryption at the level of the transport library it would be nice if we could say that endpoints always contain a public key and connections are always encrypted.

> I think encryption at very lowest level is good as we can encrypt any protocol changes over tcp / udp (such as utp) hiding even the fact we use udt / utp etc.

Doing encryption under utp is an interesting idea but it's definitely outside the scope of this RFC.

> So on listeners where do do not know if it is bootstrap (so one step) or an actual node connect we may end up with several connections per node again. Whereas on rendezvous we would not parse early messages or incorrect peers which is then single connection per peers ?

It's easy to tell why the peer is connecting, they'll say so in the first packet they send (like what we do at the moment). This is how I see it working: when a peer connects to a static listener they send a packet, encrypted with the listener's public key, containing a connection id and their public key. If the listener recognizes the connection id as one they generated rendezvous info for then they know it's a rendezvous connect, otherwise it's a direct connect. And either way having connection ids avoids us from being able to accept the same connection twice.

**Comment by dirvine** [comment]
> Why do we need unencrypted connections? We could always distribute the public keys of the seed nodes along with their addresses.
I agree. the issue is that these seed nodes will need to be "different" i.e. their config will need to hold their public and secret key. So that is possibly an addition, but not catered for yet. When the config see's these keys then it needs to sidestep key generation and use these instead. It can be done, but needs security audit and these nodes may again be untrustworthy as their keys can be stolen. So this is a swings and roundabouts one really. 
>Doing encryption under utp is an interesting idea but it's definitely outside the scope of this RFC.
Also agree, but if utp etc. is eventually integrated in connections as I think it may (if mio etc.) then this paves that way to be able to do that. 
>  If the listener recognizes the connection id
May be the issue, if the sender can send "intent" then we know what it wants. also if we get the connection before our API tells us it is expected, then perhaps we should be dropping/ignoring that connection (prob ignoring) and only report new connection when it's valid. i.e. if I want to connect as a node or relay connection then it must be agreed both sides, I should not be able to be unilateral in that respect. secure_serialisation would force that as we simply would not be able to parse until we had the info available to us.

**Comment by canndrew** [comment]
> May be the issue, if the sender can send "intent" then we know what it wants. also if we get the connection before our API tells us it is expected, then perhaps we should be dropping/ignoring that connection (prob ignoring) and only report new connection when it's valid.

This wouldn't be hard to manage. Make it so rendezvous-connecting peer uses the connection id from the `PubRendezvousInfo` of the peer they're connecting to. That way, when a `ListenerSet` receives an incoming connection it always knows whether this is for a rendezvous connect that it generated info for, even if it hasn't had `rendezvous_connect` called on it's side yet. It can just keep the connection in a cache until the user calls `rendezvous_connect`.
